Packs/ArcherRSA/Integrations/ArcherV2/ArcherV2.py

```diff
@@ -360,26 +388,51 @@ def get_token(self):
         }
         res = self._http_request('POST', 'ws/general.asmx',
                                  headers=headers, data=body, resp_type='content')
-        return extract_from_xml(
+        token = extract_from_xml(
             res,
             f'Envelope.Body.{endpoint}Response.'
             f'{endpoint}Result'
         )
+        merge_integration_context({'token': token})
+        return token
+
+    # def destroy_token(self, token):
+    #     body = terminate_session_soap_request(token)
+    #     headers = {'SOAPAction': 'http://archer-tech.com/webservices/TerminateSession',
+    #                'Content-Type': 'text/xml; charset=utf-8'}
+    #     self._http_request('POST', 'ws/general.asmx', headers=headers, data=body, resp_type='content')
+
+    def update_body_with_token(self, request_body_builder_function, create_new_token: bool = False, **kwargs):
+        time.sleep(random.uniform(0, 5))
+        context_token = get_integration_context().get('token')
+        if create_new_token or not context_token:
+            token = self.generate_token()
+        else:
+            token = context_token
+        body = request_body_builder_function(token, **kwargs)
+        return body
 
-    def destroy_token(self, token):
-        body = terminate_session_soap_request(token)
-        headers = {'SOAPAction': 'http://archer-tech.com/webservices/TerminateSession',
-                   'Content-Type': 'text/xml; charset=utf-8'}
-        self._http_request('POST', 'ws/general.asmx', headers=headers, data=body, resp_type='content')
+    def try_soap_request(self, req_data, method, create_new_token=False, attempts=1, **kwargs):
+        headers = {'SOAPAction': req_data['soapAction'], 'Content-Type': 'text/xml; charset=utf-8'}
+        request_body_builder_function = req_data['soapBody']
+        url_suffix = req_data['urlSuffix']
+        for _ in range(attempts):
+            body = self.update_body_with_token(request_body_builder_function=request_body_builder_function,
+                                               create_new_token=create_new_token, **kwargs)
+            res = self._http_request(method=method, url_suffix=url_suffix, headers=headers, data=body,
+                                     resp_type='response', ok_codes=(200, 500))
```

Comment:
in line 432 you allow 401 - should it be allowed here as well?

Reply:
Deleted from 432 (which is now 496)